crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -324,6 +324,76 @@ def services
       end
       handle_asynchronously :services
 
+      def openstackbackup
+        dump_location = "/var/lib/crowbar/upgrade/6-to-7-openstack_dump.sql"
+        if File.exist?(dump_location)
+          Rails.logger.debug("OpenStack backup already exists. Skipping...")
+          return
+        end
+
+        db_node = ::Node.find("roles:database-config-default").first
+        user = "postgres"
+        pass = db_node[:postgresql][:password][:postgres]
+        host = db_node[:postgresql].config[:listen_addresses]
+        query = "SELECT SUM(pg_database_size(pg_database.datname)) FROM pg_database;"
+        cmd = "PGPASSWORD=#{pass} psql -t -h #{host} -U #{user} -c '#{query}'"
+
+        Rails.logger.debug("Checking size of OpenStack database")
+        db_size = run_cmd(cmd)
+        unless db_size[:exit_code].zero?
+          Rails.logger.error(
+            "Failed to check size of OpenStack database: #{db_dump[:stdout_and_stderr]}"
+          )
+          raise ::Crowbar::Error::UpgradeDatabaseSizeError.new(
+            db_size[:stdout_and_stderr]
+          )
+        end
+
+        free_space = run_cmd("df -B1 -l --output='avail' / | tail -n1")
+        unless free_space[:exit_code].zero?
+          Rails.logger.error("Cannot determine free disk space: #{free_space[:stdout_and_stderr]}")
+          raise ::Crowbar::Error::UpgradeFreeDiskSpaceError.new(
+            free_space[:stdout_and_stderr]
+          )
+        end
+        if free_space[:stdout_and_stderr].strip.to_i < db_size[:stdout_and_stderr].strip.to_i
+          Rails.logger.error("Not enough free disk space to create the OpenStack database dump")
+          raise ::Crowbar::Error::UpgradeNotEnoughDiskSpaceError.new
+        end
+
+        Rails.logger.debug("Creating OpenStack database dump")
+        db_dump = run_cmd("PGPASSWORD=#{pass} pg_dumpall -h #{host} -U #{user} > #{dump_location}")
+        unless db_dump[:exit_code].zero?
```

Comment:
Is it possible that some broken dump is created in this case? If so, we should probably try to delete it...